File: src/sort/insertion-sort.ts
export const insertionSort = function<T> (data: T[]): T[] {
  for (let i = 1; i < data.length; i++) {
    for (let j = i; j > 0; j--) {
      if (data[j] < data[j - 1]) {
        const temp    = data[j - 1]
        data  [j - 1] = data[j]
        data  [j]     = temp
      } else {
        break
      }
    }
  }
  return data
}

export const insertionSortOptimization = function<T> (data: T[]): T[] {
  for (let i = 1; i < data.length; i++) {
    let value = data[i]
    let   j     = i - 1
    for (; j >= 0; j--) {
      if (data[j] > value) {
        data[j + 1] = data[j]
      } else {
        break
      }
    }
    data[j + 1] = value
  }
  return data 
}

File: src/sort/insertion-sort/c-language/insertion-sort.c
#include<stdio.h>
int *insertionSort(int *arr, int size) {
  for (int j = 1; j < size; j++) {
    for (int i = j;i >= 0; i--) {
      if (arr[i - 1] > arr[i]) {
        int temp = arr[i - 1];
        arr[i - 1] = arr[i];
        arr[i] = temp;
      } else {
        break;
      }
    }
  }
  return arr;
}

int main() {
  int arr[] = {5, 4, 3, 2, 1};
  insertionSort(arr, 5);
  for (int i = 0; i < 5; i++) {
    printf("%d ", arr[i]);
  }
  return 0;
}